### PR TITLE
refactor: warm cache

### PR DESCRIPTION
**Pull request by xgui3783**

### CODE REVIEW

File: siibra/retrieval/cache.py [hunk 1]
```diff
@@ -137,3 +141,76 @@ def build_filename(self, str_rep: str, suffix=None):
...
 
 CACHE = Cache.instance()
+
+
+class WarmupLevel(int, Enum):
+    INSTANCE = 1
+    DATA = 5
+
+
+class WarmupRegException(Exception): pass
```
Comment (AhmetNSimsek) [inline siibra/retrieval/cache.py:151]: Could you carry this to `siibra.exceptions`?

File: siibra/retrieval/cache.py [hunk 2]
```diff
@@ -137,3 +141,76 @@ def build_filename(self, str_rep: str, suffix=None):
...
+
+
+class WarmupParam(NamedTuple):
```
Comment (AhmetNSimsek) [inline siibra/retrieval/cache.py:154]: Could this class exist under `Warmup`?

File: siibra/retrieval/cache.py [hunk 3]
```diff
@@ -137,3 +141,76 @@ def build_filename(self, str_rep: str, suffix=None):
...
+
+    @staticmethod
+    def fn_eql(wrapped_fn, original_fn):
+        return wrapped_fn is original_fn or wrapped_fn.__wrapped__ is original_fn
+
+    @classmethod
+    def is_registered(cls, fn):
+        return len([warmup_fn.fn
+                    for warmup_fn in cls._warmup_fns
+                    if cls.fn_eql(warmup_fn.fn, fn) ]) > 0
```
Comment (AhmetNSimsek) [inline siibra/retrieval/cache.py:172]: I believe this is not Flake8 compatabile
```python
def is_registered(cls, fn):
    return len([
        warmup_fn.fn
        for warmup_fn in cls._warmup_fns
        if cls.fn_eql(warmup_fn.fn, fn)
    ]) > 0
```

File: siibra/volumes/parcellationmap.py [hunk 4]
```diff
@@ -1154,6 +1154,8 @@ def _assign_volume(
         return assignments
 
 
+cache.Warmup.register_warmup_fn()(Map.registry)
```
Comment (AhmetNSimsek) [inline siibra/volumes/parcellationmap.py:1157]: `Map` is an `AtlasConcept`. I am not sure this is necessary

**Review by AhmetNSimsek — COMMENTED** [review]
A few notes but otherwise LGTM

File: siibra/retrieval/cache.py (hunk 2 above)
Comment (xgui3783) [reply siibra/retrieval/cache.py:154]: messes up the referencing.

I can make this private?

Comment (AhmetNSimsek) [reply siibra/retrieval/cache.py:154]: Let us leave like this for now then. @dickscheid has a design request on such small classes but we don't have a solution yet.

File: siibra/retrieval/cache.py (hunk 3 above)
Comment (xgui3783) [reply siibra/retrieval/cache.py:172]: will fix flake and properly add it to the test

Comment (xgui3783) [reply siibra/retrieval/cache.py]: perhaps both are PEP valid? since it's flake passed.

**Comment by xgui3783** [comment]
@AhmetNSimsek can you take a look again when you get a chance?

I fixed the tests, and properly added flake

**Pull request merged**